### DISCUSSION TITLE
Will AGE be backwards compatible for future releases?

### DISCUSSION THREAD

**Discussion by rburton** [apache/age#74] [General] [answered]
I know this is a tough one to answer, but I'm wondering what will the upgrade path be for this project and if there any breaking changes in how the data is stored, will there be a migration path?

I know it's still early stages, but I am interested in using this extension for a project and I'm just curious about how things will progress over time and if there will be breaking changes that might prevent a project using an older version from upgrading.

**Comment by jrgemignani** [comment, marked as answer]
The underlying storage is through PostgreSQL, which is why agtype was based
off of jsonb. So, that shouldn't change much. If anything, we would add to
it.

Our plan is to adhere to the openCypher specification as closely as
possible, so there shouldn't be anything that works in earlier versions
that is left out in later versions.

Hopefully that answers your questions.

John

**Reply by rburton** [reply]
Thank you! This is what I assumed as well and in terms of queries, the abstraction is the cypher specification. 

I was also trying to find a feature compatibility for Cypher. Does one exist? In the past,  I was using the Roadmap to get a sense of compatibility, but there wasn't a good definitive list.

Keep up the amazing work!

**Reply by audiBookning** [reply]
I am in the same situation.

As for being able to know the working queries, the tests in the [regress](https://github.com/apache/incubator-age/tree/master/regress) folder of the repo are a great help, but they are not really sufficient to know exactly in an "easy manner" what is not yet possible to do.

Probably a good solution would be to create some tests for all the openCypher features and generate a table with all the results after any commit? It would be easy to know at a glance the compatibility with AGE?

**Comment by jrgemignani** [comment]
We try to follow -

https://www.opencypher.org/resources
https://s3.amazonaws.com/artifacts.opencypher.org/openCypher9.pdf

However, it can leave a lot to be desired at times. We also try to
reconcile against Neo4j.

John

**Comment by jrgemignani** [comment]
Point taken, I will mention that [a table of done and to do] to the team.

John

**Reply by rburton** [reply]
I also noticed that `MERGE` was in v0.3.0 and I'm guessing that https://hub.docker.com/r/sorrell/agensgraph-extension is outdated because it doesn't accept the MERGE command. I'm only using the docker container for testing out designs.

**Comment by jrgemignani** [comment]
We have releases roughly every 2 months. Those releases contain everything
up to that point in time, packed up (Docker for example) for more general
usage.

If you want the latest - and potentially bleeding edge - you can just clone
the current master and build the extension.

John

**Reply by rburton** [reply]
Okay, I did that a few minutes ago and received the same error. I searched the code base and looks like `MERGE` isn't anywhere to be found. I don't see it listed in the cypher_keywords.c https://github.com/apache/incubator-age/blob/83426d3a92bd5fee06376a38542d35efb1e6397b/src/backend/parser/cypher_keywords.c although the roadmap says it was in version 0.3.0.

The git history doesn't show it ever existed in this file, so I suspect it wasn't implemented?

**Comment by jrgemignani** [comment]
MERGE is still being worked on - it is one of the more difficult operations.

john
